pandas/indexes/base.py

```diff
@@ -2896,27 +2896,78 @@ def join(self, other, how='left', level=None, return_indexers=False):
 
     def _join_multi(self, other, how, return_indexers=True):
         from .multi import MultiIndex
-        self_is_mi = isinstance(self, MultiIndex)
-        other_is_mi = isinstance(other, MultiIndex)
 
+        def _complete_join(new_lvls, new_lbls, new_nms):
```

Comment:
first thing is that you must have exactly the same levels that match, otherwise this is a raise. They must match in number and name (or None matches None).